### PR TITLE
Church behavior

### PR DESCRIPTION
**Pull request by Anushree-naik**
Added methods in Church implementation 
1. To build pilgrims
2. To record positions of friendly churches
3. To record mirrored positions of churches 
4. To find resources in the range

The church unit is not getting invoked by our bot. This code is just to provide some implementation.

### CODE REVIEW

File: bots/psuteam7bot/castle.js [hunk 1]
```diff
@@ -264,6 +268,13 @@ castle.findPosition = (self) => {
             return 1;
         }
     });
+    self.teamCastles.sort((a,b) => {
+        if(movement.getDistance(self.me, a) > movement.getDistance(self.me, b)) {
```
Comment (lbiv4) [inline bots/psuteam7bot/castle.js]: I think this okay, but I vaguely recall that it was better to sort by id (to ensure that `self.teamCastles[0]` is yourself). Can't remember for sure tho

File: bots/psuteam7bot/church.js [hunk 2]
```diff
@@ -1,30 +1,107 @@
...
+church.maxKarbonite = SPECS.UNITS[SPECS.CHURCH].KARBONITE_CAPACITY;
+church.maxFuel = SPECS.UNITS[SPECS.CHURCH].FUEL_CAPACITY;
 
 church.doAction = (self) => {
-    self.log("church " + self.id + " taking turn.");
-    return;
+     self.log("church" + self.id + "taking turn.");
+
+    if(self.me.turn == 1){
+        const karboniteDepots = movement.getResourcesInRange(self.me, 36, self.karbonite_map);
```
Comment (lbiv4) [inline bots/psuteam7bot/church.js]: Good tweak - better to have a wider range for churches than castles

File: bots/psuteam7bot/church.js [hunk 3]
```diff
@@ -1,30 +1,107 @@
...
+        const fuelDepots = movement.getResourcesInRange(self.me, 16, self.fuel_map)
+        fuelDepots.forEach(depot => {
+            self.castleBuildQueue.push({unit: "PILGRIM", x: depot.x, y: depot.y});
+        })
+
+        const mirrorCastle = movement.getMirrorCastle(self.me, self.map)
+        self.target = mirrorCastle;
+        self.log(self.castleBuildQueue);
+
+        return church.castle.buildFromQueue(self);
```
Comment (lbiv4) [inline bots/psuteam7bot/church.js]: Just `castle.buildFromQueue` I think, You can reuse that method from `castle` just fine

File: bots/psuteam7bot/church.js [hunk 4]
```diff
@@ -1,30 +1,107 @@
...
+    }  
+    else if (self.me.turn <= 4) {
+        self.log("BUILD QUEUE NON-EMPTY")
+        self.log(self.castleBuildQueue)
+        const botsInQueue = self.castleBuildQueue.length;
+        //Keep queue at reasonable size, adding another crusader as necessary so crusaders are continually build
+        if (botsInQueue <= 5) {
+            self.castleBuildQueue.push({unit: "PROPHETS", x: self.target.x, y: self.target.y});
+        }
+        return church.castle.buildFromQueue(self);
```
Comment (lbiv4) [inline bots/psuteam7bot/church.js]: `castle.buildFromQueue(self)`

File: bots/psuteam7bot/church.js [hunk 5]
```diff
@@ -1,30 +1,107 @@
...
+ * Input: self = this is the reference to the object to the calling method. 
+ * Output: returnPosition = return value containing the positions of the friendly castle       
+ *  */
+church.recordPosition = (self) => {
+    let turn = self.me.turn;
+    if(turn <= 2){
+        self.signal(self.me.x,(self.signal_radius)^2);
+    }
+    else if(turn <= 4){
+        self.signal(self.me.y,(self.signal_radius)^2);
```
Comment (lbiv4) [inline bots/psuteam7bot/church.js]: Probably could use `castle_talk`, but I suppose `signal` might be good for visibility to other churches? Hadn't thought hard enough about that...

**Review by lbiv4 — APPROVED** [review]
There are two typos that should probably be fixed, but otherwise looks good for the end of the project

**Comment by lbiv4** [comment]
PS - feel free to ignore the "some checks were not successful message". It's just coveralls/test coverage and it's not that big of a dip to really care about

File: bots/psuteam7bot/church.js (hunk 5 above)
Comment (Anushree-naik) [reply bots/psuteam7bot/church.js]: I am not sure if castle_talk would work here or not. I thought if communication in between units. So used `(self.signal_radius)^2`

File: bots/psuteam7bot/church.js (hunk 4 above)
Comment (Anushree-naik) [reply bots/psuteam7bot/church.js]: okay

**Pull request merged**